Hermes/appscale/hermes/stats/client/remote_profiler.py

```diff
@@ -0,0 +1,199 @@
+  NodesProfileLog, ProcessesProfileLog, ProxiesProfileLog
+)
+
+INCLUDE_LISTS = IncludeLists({
+  # Node stats
+  'node': ['utc_timestamp', 'cpu', 'memory',
+           'partitions_dict', 'loadavg'],
+  'node.cpu': ['percent', 'count'],
+  'node.memory': ['available', 'total'],
+  'node.partition': ['free', 'used'],
+  'node.loadavg': ['last_5min'],
+  # Processes stats
+  'process': ['monit_name', 'unified_service_name', 'application_id',
+              'port', 'cpu', 'memory', 'children_stats_sum'],
+  'process.cpu': ['user', 'system', 'percent'],
+  'process.memory': ['resident', 'virtual', 'unique'],
+  'process.children_stats_sum': ['cpu', 'memory'],
+  # Proxies stats
+  'proxy': ['name', 'unified_service_name', 'application_id',
+            'frontend', 'backend', 'servers_count'],
+  'proxy.frontend': ['bin', 'bout', 'scur', 'smax', 'rate',
+                     'req_rate', 'req_tot', 'hrsp_4xx', 'hrsp_5xx'],
+  'proxy.backend': ['qcur', 'scur', 'hrsp_5xx', 'qtime', 'rtime'],
+})
+
+APPSCALE_PORT = 17441
+
+requests.packages.urllib3.disable_warnings()
+
+
+class RemoteProfiler(object):
+  def __init__(self, head_node_ip, secret, include_lists, output_dir,
+               verbose_processes_log, verbose_proxies_log):
+    self.head_node_ip = head_node_ip
+    self.secret = secret
+    self.nodes_log = NodesProfileLog(include_lists, output_dir)
+    self.processes_log = ProcessesProfileLog(
+      include_lists, output_dir, verbose_processes_log)
+    self.proxies_log = ProxiesProfileLog(
+      include_lists, output_dir, verbose_proxies_log)
+    self.loggers_dict = {
+      'nodes': self.nodes_log,
+      'processes': self.processes_log,
+      'proxies': self.proxies_log,
+    }
+    self.models_dict = {
+      'nodes': models.NodeStatsSnapshot,
+      'processes': models.ProcessesStatsSnapshot,
+      'proxies': models.ProxiesStatsSnapshot,
+    }
+    self.running = False
+
+  def fetch_stats(self, stats_type):
+    response = requests.get(
+      'https://{ip}:{port}/stats/cluster/{stats_type}'.format(
+        ip=self.head_node_ip, port=APPSCALE_PORT, stats_type=stats_type
+      ),
+      headers={
+        'Appscale-Secret': self.secret
+      },
+      json={
+        'include_lists': INCLUDE_LISTS.asdict(),
+        'newer_than': time.mktime(datetime.now().timetuple()),
+      },
+      verify=False
+    )
+    response.raise_for_status()
+    result_dict = response.json()
+    raw_stats = result_dict['stats']
+    failures = result_dict['failures']
+    try:
+      stats_model = self.models_dict[stats_type]
+      stats = {
+        ip: converter.stats_from_dict(stats_model, snapshot)
+        for ip, snapshot in raw_stats.iteritems()
+      }
+    except TypeError as err:
+      msg = u"Can't parse stats snapshot ({})".format(err)
+      logging.error(msg)
+      stats = {ip: msg for ip in raw_stats}
+    return stats, failures
+
+  def pull_periodically(self, stats_type, period):
+    def periodical_job():
```

Comment:
I'm not understanding why this nested function definition is needed.

Reply:
You're right, it is not necessary.